### ISSUE TITLE
OWASP Top 10 Claim

### ISSUE THREAD

**Issue by digininja** [akto-api-security/akto#14]
Your readme states:

> Akto offers tests for all OWASP top 10

Can you explain how you test for these two issues:

[A04:2021-Insecure Design](https://owasp.org/Top10/A04_2021-Insecure_Design/)
[A09:2021-Security Logging and Monitoring Failures](https://owasp.org/Top10/A09_2021-Security_Logging_and_Monitoring_Failures/)

**Comment by Ankita28g** [comment]
Hi @digininja 

Thanks for the question. 

1. We offer tests for [OWASP Top 10 for API security](https://github.com/akto-api-security/tests-library). 
2. Couple of categories are in beta. Our [roadmap is public](https://github.com/orgs/akto-api-security/projects/8/views/1). We will have it available for all very soon. 

Your suggestions are welcome.

**Comment by digininja** [comment]
Can you give a quick summary of how your beta version covers these two issues:

[API9:2019 Improper Assets Management](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa9-improper-assets-management.md)

[API10:2019 Insufficient Logging & Monitoring](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xaa-insufficient-logging-monitoring.md)

**Comment by ankush-jain-akto** [comment]
Hi @digininja 

**API9:2019 Improper Assets Management** - Old API versions: 
In many cases, we detect old API versions. For example, if an API looks like ```GET /api/v3/users```, we try to check vulnerabilities for  ```GET /api/v2/users``` and  ```GET /api/v1/users``` as well. This is solved through a test case [Old API Versions Test](https://github.com/akto-api-security/tests-library/blob/master/BOLA/business-logic/OldApiVersionTest.java).  As a future implementation, we are also considering to get list of APIs from the source code as well. This implementation is currently under very early stages of development. 

**API10:2019 Insufficient Logging & Monitoring** - 
Akto captures *all* the API traffic that comes to the application servers. User can choose to save all this traffic. This way, we can completely log all the requests. When we detect any misconfiguration (for example, endpoint sharing any sensitive info), a Slack alert also goes to the team.

**Comment by digininja** [comment]
So for 10, you are logging your requests, not testing if they are logging
anything, is that right?

And if you find an issue, you are doing the alerting, you aren't testing if
they are sending off any kind of alerts internally.

**Comment by ankush-jain-akto** [comment]
Yes. We are logging it ourselves. User can send to any logging tool as well (eg DataDog, Newrelic etc.) We are *not testing* if they are being logged already. 

We are not testing if requests are logged on a diff tool or if alerts are set on a diff tool. Users can choose Akto to log & alert or send this data to their favourite tool and set logging/alerting from there itself.

**Comment by digininja** [comment]
So you aren't testing either of these top 10 issues. I think you need to
look at the wording of your readme as it is misleading at the moment.

**Comment by ankush-jain-akto** [comment]
Hi @digininja - we do solve for API9:2019 Improper Assets Management through a test case [Old API Versions Test](https://github.com/akto-api-security/tests-library/blob/master/BOLA/business-logic/OldApiVersionTest.java) as I mentioned above. 

I will update the Readme file accordingly. I will make a PR linked to this issue too.

**Comment by digininja** [comment]
Looking to see if there is a v2 when you are testing v3 isn't really doing
any kind of thorough testing. If a tester told me that that is all they did
to look for that area then I'd send them back to do a lot more work.

**Comment by digininja** [comment]
You still say you cover all the top 10 issues which you don't do.

**Comment by ankush-jain-akto** [comment]
Hi @digininja,
These discussions are quite useful to us. Really appreciate you helping us fill the understanding gaps. We want to be very transparent & provide complete info about what we exactly do. Thanks from our entire team for this! I am happy to say we are yet to cover a particular category, if that is really the case. I want to be thorough to understand if this is a documentation problem or actually insignificant coverage problem.
I am assuming API-9- Improper Assets Management is the subject of concern here. Taking a section from [OWASP website about this issue](https://github.com/OWASP/API-Security/blob/master/2019/en/src/0xa9-improper-assets-management.md#is-the-api-vulnerable), I am adding how Akto helps in each of these -

1. The purpose of an API host is unclear, and there are no explicit answers to the following questions:
    1. _Which environment is the API running in (e.g., production, staging, test, development)?_ Akto provides this out of the box. If you install Akto using Traffic Mirroring in your VPC, this happens automatically in real time.
    2. _Who should have network access to the API (e.g., public, internal, partners)?_ For each API on Akto dashboard, you can see a column called "Access Type" which says Public/Internal.
    3. _Which API version is running?_ If API version is a part of the URL, this again is logged in Akto dashboard and users can tag the APIs containing "v1" as #version-1. They can also set alerts - if a new v1 API is added (no additions should be done to old API versions)
    4. _What data is gathered and processed by the API (e.g., PII)?_ Akto provides this too out of the box. If some PII patterns are very business specifc, users can add their custom PII patterns too.
    5. _What's the data flow?_ This isn't covered yet.
2. _There is no documentation, or the existing documentation is not updated_ - Akto prepares an inventory for all APIs across all services. It also provides a way for users to download it as an OpenAPI spec file or export it to their Postman workspace. Users can do it any number of times. In fact, they can prepare a cron job to extract OpenAPI spec file from Akto and save it in their repo.
3. _There is no retirement plan for each API version_ - This is more of a process fix rather than a product fix. But Akto also mentions APIs that are not being used anymore. Users can take these APIs and test themselves if they are running. They can also export it as a CSV and send it to dev teams to confirm if they are shut down from backend also.
4. _Hosts inventory is missing or outdated_ - Akto can fill this too. Because it captures all the traffic, all the hosts are also automatically registered along with all their APIs too.
5. _Integrated services inventory, either first- or third-party, is missing or outdated_ - Akto doesn't provide this out of the box. This requires an integration. Third party services are typically called via https. We get this traffic too for analysis, but nothing useful comes out because its encrypted. However, devs can use our SDKs to send a copy of the HTTP call to Akto and then Akto dashboard can show the 3rd party calls as well. Too much work - meant for desperate users only I would say.
6. _Old or previous API versions are running unpatched_ - If old versions are running, they will be logged by Akto and tested too for all vuln.

You would see a lot of dependency on traffic here - which is true for Akto. Akto's source of information is traffic data. As long as there is even 1 API hit, Akto will log it and analyze it for vulnerabilities. Akto will also document it and users can check for its traffic usage, download openapi-spec, set alerts etc.
If an API is never hit, it won't be logged by Akto. In such cases, we do try to find these inactive APIs by changing version numbers. In future we want to launch an endpoint discovery module too - through fuzzing (sitting outside the application) or code analysis (sitting inside the application).
Does this help @digininja? Please let me know if we are missing major cases. Also, putting in a user-research plug here. Personally, I know you are a legend when it comes to this space. Would love to know more from you on how you envision some of the API security vuln be exposed through testing. Your suggestions will be quite valuable to us & I promise I will put them in our product.

**Comment by digininja** [comment]
My problem is with this statement:

>  Akto offers coverage for all OWASP top 10 and HackerOne Top 10 categories

You don't test for API:10, you can't do it in an automated tool and so your statement is wrong.

For API:9, you may cover it, but I would argue that it can't be automated, you may be able to give some clues to things which may be there when they shouldn't, but you can't state there are any issues as the tool doesn't know enough about an environment to make that call. It may be that you find versions 2 and 3 but they are both designed to be live.

Asset management is also a lot more than just a version number.

Have you considered mapping yourself to the https://owasp.org/www-project-web-security-testing-guide/ instead as that is designed to be testable whereas the Top 10 is just a list of common issues, it isn't designed to be something you can automate or say that you cover.

**Comment by craig-shony** [comment]
I wonder if the appropriate wording would be "Akto offers coverage for all applicable OWASP Top 10 (API Security)" because it's not possible for a DAST-like tool such as Akto to provide coverage for all 10. As OP mentioned, some of the Top 10 are not suitable for automated testing.

**Comment by digininja** [comment]
That would be more appropriate.

**Comment by Ankita28g** [comment]
hi @digininja @craig-shony 

Acknowledged. We will be keeping this issue open and get more feedback. Meanwhile, we will also be getting more first hand opinions from folks in our network.

will be update the wording here once we collect feedback.

thanks!
